src/portable/ehci/ehci.c

```diff
@@ -457,6 +470,9 @@ bool hcd_edpt_xfer(uint8_t rhport, uint8_t dev_addr, uint8_t ep_addr, uint8_t *
     // attach head QTD to QHD start transferring
     p_qhd->qtd_overlay.next.address = (uint32_t) p_qhd->p_qtd_list_head;
   }
+  // Flush data cache if required, see hcd.h for details
+  hcd_dcache_flush((uint32_t)p_qhd, sizeof(ehci_qhd_t));
```

Comment:
we probably need to also do invalidate + flush since EHCI does write back to update qhd/qtd as well ? or we could do the invalidate in the hcd_int_handler() I guess.

Reply:
Personally, I wasn't keen on using cache functions inside the interrupt handler as it adds latency to the interrupt, on my platform that might be around 6uS worst case which isn't hideous but it will vary on other platforms.  Do we want to try it this way or try and keep the cache code away from the ISR?  

What do you think? 

Thanks for your feedback @hathach.

Kind Regards,

Pete

Reply:
sure we can skip dcache or any function within ISR if possible. I think you could try to both invalidate and flush following memory location in hcd_edpt_xfer() an hopefully this would be enough:
- data buffer
- qtd (since ehci controller update this)
- qhd (ehci also update it via qtd overlay).